server/src/main/java/com/cloud/usage/UsageServiceImpl.java

```diff
@@ -216,6 +216,31 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
             s_logger.debug("Account details not available. Using userContext accountId: " + accountId);
         }
 
+        // Check if a domain admin is allowed to access the requested account info.
+        if (_accountService.isDomainAdmin(caller.getId()) && accountId != null){
+            long accountDomainId = _accountDao.getDomainIdForGivenAccountId(accountId);
+            long callerDomainId = caller.getDomainId();
+            boolean matchFound = false;
+
+            if (callerDomainId == accountDomainId) {
+                matchFound = true;
+            } else {
+                // Check if the account is in a child domain of this domain admin.
+                List<DomainVO> childDomains = _domainDao.findAllChildren(_domainDao.findById(caller.getDomainId()).getPath(), caller.getDomainId());
+
+                for (DomainVO domainVO: childDomains) {
+                    if (accountDomainId == domainVO.getId()) {
```

Comment:
Why not use the existing `_accountMgr.buildACLSearchParameters`?
And should the be done for all caller types, as: (this is just suggestion, I've not investigated this)
```
Root admin - allow recursive usage records for all domains
Domain admin - only allow recursive usage records for its own domain
User - don't allow recursive domain ID (it should be only allowed to see their own usage records)
```

Reply:
The _accountMgr.buildACLSearchParameter method needs a seaarchBuilder and not searchCriteria. I believe you intended it for fetching usage records, but I'm I am simply doing authorization here. 

If a Root admin does the call with a domain id, that is handled on line 276. If a domain admin does the call, authorization will happen first. Users cannot call this API.

Reply:
@Spaceman1984 can we use checkAccess method to simplify checking accountId domain access?
`void checkAccess(Account account, Domain domain) throws PermissionDeniedException;`

Reply:
Addressed